NvmeOfDxe+SpdkShim integration with SPDK LTS v23.01.1

trevor-cockrell opened the pull request:
Full adjustment to NvmeOf driver+shim to separate from any required SPDK-source changes. Based on SPDK LTS v23.01.1.

### Please note: Due to compilation dependencies between our driver, shim layer, and the submodule, this is not organized into "EDK2-ified" commits. A final rebase will be required once approved.
For now, commits are organized by general section for review*, with the intent being to rebase this merged code into a final 'release-branch' for EDK2 staging.

*Due to the lift of this effort, some commits are much heavier than others, notably: 
- 9d6bd5794e76a16e174de84c3d8598a9a6c31483: This introduces all required changes to the shim-layer alongside ported spdk code.
  - This adds support for EDK2-required context within the SPDK layer, involving several imported SPDK functions which either required modifications or were required for compilation
- 0a592c83c70581edec6962dfb5c79ee6d1df2040: This introduces all imported SPDK source required for compilation. Details can be found in [section 4.14.2](https://github.com/timberland-sig/edk2/blob/c4b9242ca74ef826f46c84c01582ef6c2dd7bd07/NetworkPkg/NvmeOfDxe/readme.md#4142-spdk-importing)  of the readme
 

Since the SPDK submodule has been changed, please ensure that it is updated before attempting to build this branch.

----

Addtl. notes:
- Most changes involved porting and importing SPDK code to support our function calls and enable MSVC compilation
  - SPDK source importing attempted to be as minimal as possible. Details are noted in the [readme](https://github.com/timberland-sig/edk2/blob/c4b9242ca74ef826f46c84c01582ef6c2dd7bd07/NetworkPkg/NvmeOfDxe/readme.md) (c4b9242ca74ef826f46c84c01582ef6c2dd7bd07)
- Built and tested with both GCC and MSVC
- Resolved all GCC-introduced Coverity issues (NvmeOfDxe, SpdkShim)
  - 15 Coverity issues found in the spdk submodule during GCC build. These cannot be resolved without pulling more SPDK source into our shim.
  - ! No progress on resolving Coverity issues introduced via MSVC build

Patch checker passes (again, a rebase is still required, but the diffs at least pass):
```
$ patchcheck bf4424c7~1..HEAD --one
[DEV]Documentation updates
c4b9242ca7 ok
[DEV]NetworkPkg/Application/NvmeOfCli
c05ced11e8 ok
[DEV]NetworkPkg/NvmeOfDxe
a9e7dbdf04 ok
[DEV]NetworkPkg/Library/DxeSpdkLib
8432291799 ok
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim
9d6bd5794e ok
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim/sys
50549a55ef ok
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim/spdk
0a592c83c7 ok
NetworkPkg/Library/DxeSpdkLib: Update SPDK submodule
bf4424c7fe ok
```

trevor-cockrell committed:
NetworkPkg/Library/DxeSpdkLib: Update SPDK submodule

Bump SPDK to LTS 23.01.1

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim/spdk

Update spdk with ported modules

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim/sys

Add required system headers

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/Library/DxeSpdkLib/SpdkShim

Full SpdkShim update for now

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/Library/DxeSpdkLib

Update DxeSpdkLib.inf

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/NvmeOfDxe

Update NvmeOfDxe with SPDK changes

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]NetworkPkg/Application/NvmeOfCli

Update NvmeOfCli with SPDK API calls

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell committed:
[DEV]Documentation updates

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell commented:
As noted in SIG call (11/7), will be updating our timberland-sig/spdk repository with a branch to track the changes introduced by this PR.

tbzatek commented:
FYI, while I haven't retested any of my currently open issues, I'm using this code for my daily work. Haven't spotted any regression so far, compiled fine as well. So far so good.

johnmeneghini reviewed (commented):
It looks like you are including all of the spdk files in this PR.

Is that the intention?  Do we plan to submit the actual SPDK source code, upstream, to EDK2?

trevor-cockrell committed:
[DEV]Documentation updates (SPDK)

Remove SPDK diffs in appendix. Update SPDK importing section to point to
the timberland-sig/spdk branch used to track our modified imported
source instead.

Signed-off-by: Trevor Cockrell <trevor_cockrell@dell.com>

trevor-cockrell commented:
@johnmeneghini  @mwilck 

https://github.com/timberland-sig/edk2/pull/39/commits/f7439cbc978a0720a49695a6dd09e189b97d1d22 updates the readme to point at our SIG/spdk dev branch for tracking modifications to imported SPDK source
- https://github.com/timberland-sig/spdk/tree/timberland-modified-imports

This is done with two commits on top of the LTS:
- https://github.com/timberland-sig/spdk/commit/e3381c226033933bc65e795c513ffebaa5e6386a Purely EDK2 Uncrustify formatting. No functional changes.
- https://github.com/timberland-sig/spdk/commit/60c439a189dab190174b6a8d53cfe1d889ed0019 Functional changes to the imported source files.

Please let me know if you have any further feedback or suggestions!

--
> It looks like you are including all of the spdk files in this PR.
> 
> Is that the intention? Do we plan to submit the actual SPDK source code, upstream, to EDK2?

Just putting notes here from last discussion. The only SPDK source being pulled in (and, yes, ultimately submitted to EDK2) are the files tracked in the above branch. Although not ideal, this is at least limited to these 8 modified files.

mwilck commented:
@trevor-cockrell, if you rebase this PR now we'd get ready-built OVMF images from GitHub actions...

tbzatek reviewed (changes requested):
This branch fails to build with gcc-15, on multiple places. Upstream edk2/master compiles fine, but their changes cannot be easily backported. One of which is https://github.com/tianocore/edk2/pull/10647

There's more, e.g.
```
/tmp/edk2/NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c: In function ‘NvmeOfCliProbeControllers’:
/tmp/edk2/NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:1444:9: error: passing argument 3 of ‘spdk_nvme_probe’ from incompatible pointer type [-Wincompatible-pointer-types]
 1444 |         NvmeOfCliProbeCallback,
      |         ^~~~~~~~~~~~~~~~~~~~~~
      |         |
      |         BOOLEAN (*)(void *, const struct spdk_nvme_transport_id *, struct spdk_nvme_ctrlr_opts *) {aka unsigned char (*)(void *, const struct spdk_nvme_transport_id *, struct spdk_nvme_ctrlr_opts *)}
In file included from /tmp/edk2/NetworkPkg/Library/DxeSpdkLib/spdk/lib/nvme/nvme_internal.h:14,
                 from /tmp/edk2/NetworkPkg/Library/DxeSpdkLib/SpdkShim/edk_nvme_ctrlr.h:12,
                 from /tmp/edk2/NetworkPkg/NvmeOfDxe/NvmeOfDriver.h:44,
                 from /tmp/edk2/NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:10:
/tmp/edk2/NetworkPkg/Library/DxeSpdkLib/spdk/include/spdk/nvme.h:874:40: note: expected ‘spdk_nvme_probe_cb’ {aka ‘_Bool (*)(void *, const struct spdk_nvme_transport_id *, struct spdk_nvme_ctrlr_opts *)’} but argument is of type ‘BOOLEAN (*)(void *, const struct spdk_nvme_transport_id *, struct spdk_nvme_ctrlr_opts *)’ {aka ‘unsigned char (*)(void *, const struct spdk_nvme_transport_id *, struct spdk_nvme_ctrlr_opts *)’}
  874 |                     spdk_nvme_probe_cb probe_cb,
      |                     ~~~~~~~~~~~~~~~~~~~^~~~~~~~
/tmp/edk2/NetworkPkg/NvmeOfDxe/NvmeOfCliInterface.c:1198:1: note: ‘NvmeOfCliProbeCallback’ declared here
 1198 | NvmeOfCliProbeCallback (
      | ^~~~~~~~~~~~~~~~~~~~~~
/tmp/edk2/NetworkPkg/Library/DxeSpdkLib/spdk/include/spdk/nvme.h:789:16: note: ‘spdk_nvme_probe_cb’ declared here
  789 | typedef bool (*spdk_nvme_probe_cb)(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
      |                ^~~~~~~~~~~~~~~~~~

```

tbzatek commented:
An exquisitely dirty patch: [gcc-15.patch](https://github.com/user-attachments/files/22024420/gcc-15.patch)
It appears to boot just as fine.

mwilck commented:
> An exquisitely dirty patch: gcc-15.patch]

I wouldn't call it "dirty". It's quite fine actually. But still I'd recommend against applying it here. This is the kind of change that upstream has to take care of.

I, for one, use containers to build EDK2, so I don't have to worry about compatibility issues with the latest compilers.

johnmeneghini commented:
Both Tomas and I have tested this and I think it should be merged.

tbzatek commented:
> I wouldn't call it "dirty". It's quite fine actually. But still I'd recommend against applying it here. This is the kind of change that upstream has to take care of.

Absolutely agree, posted just as a temporary workaround. As noted on the last call, it's fixed upstream so we should rather rebase all our branches first.

tbzatek commented:
> I wouldn't call it "dirty". It's quite fine actually.

The dirtiness here is to avoid typecasting the callback to `(spdk_nvme_probe_cb)` and rather fix the actual callback arguments to conform.
